src/workerd/api/filesystem.c++

```diff
@@ -0,0 +1,442 @@
+#include "filesystem.h"
+
+#include "blob.h"
+
+#include <workerd/api/streams/standard.h>
+
+namespace workerd::api {
+
+// =======================================================================================
+// Implementation of the Web File System API
+
+namespace {
+bool isValidFileName(kj::StringPtr name) {
+  return name.size() > 0 && name != "."_kj && name != ".."_kj && name.find("/"_kj) == kj::none;
+}
+}  // namespace
+
+jsg::Promise<jsg::Ref<FileSystemDirectoryHandle>> StorageManager::getDirectory(
+    jsg::Lock& js, const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& exception) {
+  KJ_IF_SOME(vfs, workerd::VirtualFileSystem::tryGetCurrent(js)) {
+    return js.resolvedPromise(
+        js.alloc<FileSystemDirectoryHandle>(jsg::USVString(), vfs.getRoot(js)));
+  }
+
+  auto ex = js.domException(kj::str("SecurityError"), kj::str("No current virtual file system"));
+  return js.rejectedPromise<jsg::Ref<FileSystemDirectoryHandle>>(exception.wrap(js, kj::mv(ex)));
+}
+
+FileSystemDirectoryHandle::FileSystemDirectoryHandle(
+    jsg::USVString name, kj::Rc<workerd::Directory> inner)
+    : FileSystemHandle(kj::mv(name)),
+      inner(kj::mv(inner)) {}
+
+jsg::Promise<bool> FileSystemDirectoryHandle::isSameEntry(
+    jsg::Lock& js, jsg::Ref<FileSystemHandle> other) {
+  KJ_IF_SOME(other, kj::dynamicDowncastIfAvailable<FileSystemDirectoryHandle>(*other)) {
+    return js.resolvedPromise(other.inner.get() == inner.get());
+  }
+  return js.resolvedPromise(false);
+}
+
+jsg::Promise<jsg::Ref<FileSystemFileHandle>> FileSystemDirectoryHandle::getFileHandle(jsg::Lock& js,
+    jsg::USVString name,
+    jsg::Optional<FileSystemGetFileOptions> options,
+    const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& exception) {
+  JSG_REQUIRE(isValidFileName(name), TypeError, "Invalid file name");
+  kj::Maybe<FsType> createAs;
+  KJ_IF_SOME(opts, options) {
+    if (opts.create) createAs = FsType::FILE;
+  }
+  KJ_IF_SOME(node,
+      inner->tryOpen(js, kj::Path({name}),
```

Comment:
I regret the original decision to let tryOpen return `kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>, kj::Rc<SymbolicLink>>>` with the comment 
```
  // If the path identifies a symbolic link, the symbolic link will be resolved
  // and the target, if it exists, will be returned. If the target does not exist,
  // kj::none will be returned.
```
Is there any reason not to amend that to return `kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>>>` instead and skip all of the 
```
      KJ_CASE_ONEOF(link, kj::Rc<workerd::SymbolicLink>) {
        KJ_UNREACHABLE;
      }
```
These are counter-intuitive and confusing without real value.

Reply:
These are *currently* `KJ_UNREACHABLE` because we're not doing anything with symlinks yet. Once we fill out the full node:fs implementation these may not remain `KJ_UNREACHABLE`. Once the implementation is done I'll be going back and revisiting these.